### PR TITLE
[Release-2.5] Port fix from master to release-2.0.5: Serialize type alias when type-alias is not accessible and emit generic

### PR DESCRIPTION
**Pull request by yuit**
#11035
#10625

### CODE REVIEW

**Comment by mhegazy** [comment]
@sheetalkamat can you take a look

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2152,14 +2165,16 @@ namespace ts {
...
                     }
-                    else if (!(flags & TypeFormatFlags.InTypeAlias) && type.flags & (TypeFlags.Anonymous | TypeFlags.UnionOrIntersection) && type.aliasSymbol) {
-                        if (type.flags & TypeFlags.Anonymous || !(flags & TypeFormatFlags.UseTypeAliasValue)) {
-                            const typeArguments = type.aliasTypeArguments;
-                            writeSymbolTypeReference(type.aliasSymbol, typeArguments, 0, typeArguments ? typeArguments.length : 0, nextFlags);
-                        }
-                        else {
-                            writeUnionOrIntersectionType(<UnionOrIntersectionType>type, nextFlags);
-                        }
+                    else if (!(flags & TypeFormatFlags.InTypeAlias) && ((type.flags & TypeFlags.Anonymous && !(<AnonymousType>type).target) || type.flags & TypeFlags.UnionOrIntersection) && type.aliasSymbol &&
```
Comment (sheetalkamat) [inline src/compiler/checker.ts]: This condition looks too complicated. Can you make it readable by adding comments?

File: tests/baselines/reference/controlFlowBinaryOrExpression.types [hunk 2]
```diff
@@ -62,21 +62,21 @@ declare function isHTMLCollection(sourceObj: any): sourceObj is HTMLCollection;
...
->EventTargetLike : EventTargetLike
+>EventTargetLike : NodeList | HTMLCollection | { a: string; }
 >a : string
 >HTMLCollection : HTMLCollection
 >NodeList : NodeList
 
 var sourceObj: EventTargetLike = <any>undefined;
->sourceObj : EventTargetLike
->EventTargetLike : EventTargetLike
+>sourceObj : NodeList | HTMLCollection | { a: string; }
```
Comment (sheetalkamat) [inline tests/baselines/reference/controlFlowBinaryOrExpression.types]: Shouldn't this be sourceObj: EventTargetLike?

File: src/compiler/checker.ts (hunk 1 above)
Comment (yuit) [inline src/compiler/checker.ts]: there is a comment already [here](https://github.com/Microsoft/TypeScript/pull/11392/files/40b5627b77b66d0c33224bbeecd5fccec07356d8#diff-c3ed224e4daa84352f7f1abcd23e8ccaR2170)

**Pull request merged**